datadog/resource_datadog_service_level_objective.go

```diff
@@ -0,0 +1,433 @@
+				ConflictsWith: []string{"monitor_ids", "monitor_search"},
+				Description:   "The metric query of good / total events",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"numerator": {
+							Type:     schema.TypeString,
+							Required: true,
+							StateFunc: func(val interface{}) string {
+								return strings.TrimSpace(val.(string))
+							},
+						},
+						"denominator": {
+							Type:     schema.TypeString,
+							Required: true,
+							StateFunc: func(val interface{}) string {
+								return strings.TrimSpace(val.(string))
+							},
+						},
+					},
+				},
+			},
+
+			// Monitor-Based SLO
+			"monitor_ids": {
+				Type:          schema.TypeSet,
+				Optional:      true,
+				ConflictsWith: []string{"query", "monitor_search"},
+				Description:   "A static set of monitor IDs to use as part of the SLO",
+				Elem:          &schema.Schema{Type: schema.TypeInt, MinItems: 1},
+			},
+			"monitor_search": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"query", "monitor_ids"},
+				Description:   "A dynamic search on creation for the SLO",
+			},
+			"groups": {
+				Type:          schema.TypeSet,
+				Optional:      true,
+				Description:   "A static set of groups to filter monitor-based SLOs",
+				ConflictsWith: []string{"query"},
+				Elem:          &schema.Schema{Type: schema.TypeString, MinItems: 1},
+			},
+		},
+	}
+}
+
+// ValidateServiceLevelObjectiveTypeString is a ValidateFunc that ensures the SLO is of one of the supported types
+func ValidateServiceLevelObjectiveTypeString(v interface{}, k string) (ws []string, errors []error) {
+	switch v.(string) {
+	case datadog.ServiceLevelObjectiveTypeMonitor:
+		break
+	case datadog.ServiceLevelObjectiveTypeMetric:
+		break
+	default:
+		errors = append(errors, fmt.Errorf("invalid type %s specified for SLO", v.(string)))
+	}
+	return
+}
+
+func buildServiceLevelObjectiveStruct(d *schema.ResourceData) *datadog.ServiceLevelObjective {
+
+	slo := datadog.ServiceLevelObjective{
+		Type:        datadog.String(d.Get("type").(string)),
+		Name:        datadog.String(d.Get("name").(string)),
+		Description: datadog.String(d.Get("description").(string)),
+	}
+
+	if attr, ok := d.GetOk("tags"); ok {
+		tags := make([]string, 0)
+		for _, s := range attr.(*schema.Set).List() {
+			tags = append(tags, s.(string))
+		}
+		// sort to make them determinate
+		if len(tags) > 0 {
+			sort.Strings(tags)
+			slo.Tags = tags
+		}
+	}
+
+	if attr, ok := d.GetOk("thresholds"); ok {
+		sloThresholds := make(datadog.ServiceLevelObjectiveThresholds, 0)
+		for _, rawThreshold := range attr.([]interface{}) {
+			threshold := rawThreshold.(map[string]interface{})
+			t := datadog.ServiceLevelObjectiveThreshold{}
+			if tf, ok := threshold["timeframe"]; ok {
+				t.TimeFrame = datadog.String(tf.(string))
+			}
+
+			if targetValue, ok := threshold["target"]; ok {
+				t.Target = datadog.Float64(targetValue.(float64))
```

Comment:
Is it safe to type assert without checking on these optional fields?

Reply:
will check and fix